traits/trait_set_object.py

```diff
@@ -12,247 +12,425 @@
         if id_self in memo:
             return memo[id_self]
 
-        memo[id_self] = result = TraitSetObject(
-            self.trait,
-            lambda: None,
-            self.name,
+        # notifiers are transient and should not be copied
+        memo[id_self] = result = TraitSet(
             [copy.deepcopy(x, memo) for x in self],
+            validator=copy.deepcopy(self.validator, memo),
+            notifiers=[],
         )
 
         return result
 
-    def update(self, value):
-        if not hasattr(self, "trait"):
-            set.update(self, value)
-            return
+    def __getstate__(self):
+        """ Get the state of the object for serialization.
 
-        try:
-            if not isinstance(value, set):
-                value = set(value)
-            added = value.difference(self)
-            if len(added) > 0:
-                object = self.object()
-                validate = self.trait.item_trait.handler.validate
-                if validate is not None:
-                    name = self.name
-                    added = set(
-                        [validate(object, name, item) for item in added]
-                    )
-
-                set.update(self, added)
-
-                if self.name_items is not None:
-                    self._send_trait_items_event(
-                        self.name_items, TraitSetEvent(None, added)
-                    )
-        except TraitError as excp:
-            excp.set_prefix("Each element of the")
-            raise excp
+        Notifiers are transient and should not be serialized.
+        """
+        result = self.__dict__.copy()
+        # notifiers are transient and should not be serialized
+        result.pop("notifiers", None)
+        return result
 
-    def intersection_update(self, value):
-        removed = self.difference(value)
-        if len(removed) > 0:
-            set.difference_update(self, removed)
+    def __setstate__(self, state):
+        """ Restore the state of the object after serialization.
+
+        Notifiers are transient and are restored to the empty list.
+        """
+        state['notifiers'] = []
+        self.__dict__.update(state)
+
+    def add(self, value):
+        """ Add an element to a set.
+
+        This has no effect if the element is already present.
+
+        Parameters
+        ----------
+        value : any
+            The value to add to the set.
+
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                An empty set.
+            added : set
+                A set containing the added value.
 
-            if self.name_items is not None:
-                self._send_trait_items_event(
-                    self.name_items, TraitSetEvent(removed)
-                )
+        """
+        validated_values = self.validate(set([value]))
+        if len(validated_values) > 1:
+            raise ValueError("Validator returned {} values "
```

Comment:
Please could you add a unit test exercising this path? (Though if we replace the whole-set validator with an item validator, this problem goes away.)

Reply:
Addressed in #998 - No longer used.